internal/controllers/machineset/machineset_preflight.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	errList := []error{}
+	preflightCheckErrs := []preflightCheckError{}
+	// Run the control-plane-stable preflight check.
+	if !skipped.Has(clusterv1.MachineSetPreflightCheckControlPlaneStable) {
+		preflightCheckErr, err := r.controlPlaneStablePreflightCheck(controlPlane)
+		if err != nil {
+			errList = append(errList, err)
+		}
+		if preflightCheckErr != nil {
+			preflightCheckErrs = append(preflightCheckErrs, preflightCheckErr)
+		}
+	}
+
+	// Check the version skew policies only if version is defined in the MachineSet.
+	if ms.Spec.Template.Spec.Version != nil {
+		msVersion := *ms.Spec.Template.Spec.Version
+		msSemver, err := semver.ParseTolerant(msVersion)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to parse version %q of MachineSet %s", action, msVersion, klog.KObj(ms))
+		}
+
+		// Run the kubernetes-version skew preflight check.
+		if !skipped.Has(clusterv1.MachineSetPreflightCheckKubernetesVersion) {
+			preflightCheckErr := r.kubernetesVersionPreflightCheck(cpSemver, msSemver)
+			if preflightCheckErr != nil {
+				preflightCheckErrs = append(preflightCheckErrs, preflightCheckErr)
+			}
+		}
+
+		// Run the kubeadm-version skew preflight check.
+		if !skipped.Has(clusterv1.MachineSetPreflightCheckKubeadmVersion) {
+			preflightCheckErr, err := r.kubeadmVersionPreflightCheck(cpSemver, msSemver, ms)
+			if err != nil {
+				errList = append(errList, err)
+			}
+			if preflightCheckErr != nil {
+				preflightCheckErrs = append(preflightCheckErrs, preflightCheckErr)
+			}
+		}
+	}
+
+	if len(errList) > 0 {
+		return ctrl.Result{}, errors.Wrapf(kerrors.NewAggregate(errList), "failed to perform %q: failed to perform preflight checks", action)
+	}
+	if len(preflightCheckErrs) > 0 {
+		preflightCheckErrStrings := []string{}
+		for _, v := range preflightCheckErrs {
+			preflightCheckErrStrings = append(preflightCheckErrStrings, *v)
+		}
+		log.Info(fmt.Sprintf("While performing %q preflight checks failed: %s", action, strings.Join(preflightCheckErrStrings, ",")))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) controlPlaneStablePreflightCheck(controlPlane *unstructured.Unstructured) (preflightCheckError, error) {
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// Check that the control plane is not provisioning.
+	isProvisioning, err := contract.ControlPlane().IsProvisioning(controlPlane)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to perform %q preflight check: failed to check if ControlPlane %s is provisioning", clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef)
+	}
+	if isProvisioning {
+		return pointer.String(fmt.Sprintf("preflight check %q failed: ControlPlane %s is provisioning", clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef)), nil
```

Comment:
Wondering if this is the right wording, because "preflight check is failed" might suggest the user there is some issue, while instead this is a handled condition that we are handling by deferring machine creation.
(same for other preflight checks).

However, this really depend by how this condition is going to surface to the users

Reply:
But isn't it correct that the preflight check failed? 

And depending on the situation the user might have to do something, but this depends on a lot of factors (e.g. ClusterClass vs legacy cluster)

Reply:
> But isn't it correct that the preflight check failed?

This is correct from an implementation PoV, but I'm not sure this is the best approach from a user PoV.

What about something like "Creation of a new Machine is on hold while CP is provisioning"? This tells the user what he is waiting for, that the operation will restart which are IMO the things he cares about the most. We can also eventually add something like "preflight check ... failed" in between parentheses if we want as additional information, but it should be more an additional info than the core of the message

Reply:
Ah I see. Yeah that sounds reasonable.

Reply:
looks great!